codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/GoJmespathExpressionGenerator.java

```diff
@@ -0,0 +1,142 @@
+import software.amazon.smithy.jmespath.JmespathExpression;
+import software.amazon.smithy.jmespath.ast.FieldExpression;
+import software.amazon.smithy.jmespath.ast.FunctionExpression;
+import software.amazon.smithy.jmespath.ast.ProjectionExpression;
+import software.amazon.smithy.jmespath.ast.Subexpression;
+import software.amazon.smithy.model.shapes.ListShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.utils.SmithyInternalApi;
+
+/**
+ * Traverses a JMESPath expression, producing a series of statements that evaluate the entire expression. The generator
+ * is shape-aware and the return indicates the underlying shape being referenced in the final result.
+ * <br/>
+ * Note that the use of writer.write() here is deliberate, it's easier to structure the code in that way instead of
+ * trying to recursively compose/organize Writable templates.
+ */
+@SmithyInternalApi
+public class GoJmespathExpressionGenerator {
+    private final GoCodegenContext ctx;
+    private final GoWriter writer;
+    private final Shape input;
+    private final JmespathExpression root;
+
+    private int idIndex = 1;
+
+    public GoJmespathExpressionGenerator(GoCodegenContext ctx, GoWriter writer, Shape input, JmespathExpression expr) {
+        this.ctx = ctx;
+        this.writer = writer;
+        this.input = input;
+        this.root = expr;
+    }
+
+    public Result generate(String ident) {
+        writer.write("v1 := $L", ident);
+        return visit(root, input);
+    }
+
+    private Result visit(JmespathExpression expr, Shape current) {
+        if (expr instanceof FunctionExpression) {
+            return visitFunction((FunctionExpression) expr, current);
+        } else if (expr instanceof FieldExpression) {
+            return visitField((FieldExpression) expr, current);
+        } else if (expr instanceof Subexpression) {
+            return visitSub((Subexpression) expr, current);
+        } else if (expr instanceof ProjectionExpression) {
+            return visitProjection((ProjectionExpression) expr, current);
+        } else {
+            throw new CodegenException("unhandled jmespath expression " + expr.getClass().getSimpleName());
+        }
+    }
+
+    private Result visitProjection(ProjectionExpression expr, Shape current) {
+        var left = visit(expr.getLeft(), current);
+
+        // left of projection HAS to be an array by spec, otherwise something is wrong
+        if (!left.shape.isListShape()) {
+            throw new CodegenException("left side of projection did not create a list");
+        }
+
+        var leftMember = expectMember(ctx.model(), (ListShape) left.shape);
+
+        // We have to know the element type for the list that we're generating, use a dummy writer to "peek" ahead and
+        // get the traversal result
+        var lookahead = new GoJmespathExpressionGenerator(ctx, new GoWriter(""), leftMember, expr.getRight())
+                .generate("v");
+
+        ++idIndex;
```

Comment:
not quite sure I follow why we increase the index when visiting the projection. Each field increases the index because we're visiting a concrete field, but why do we increase it here?

Reply:
We need a new identifier for the slice variable that holds the result of the overall projection (it's declared in the following write()).